### PR TITLE
Mechanics: Frugal performance improvement and clarification

### PR DESCRIPTION
**Pull request by Hurleveur**
Just improved the logic a bit to not calculate stuff without need.
Also I made sure to clarify the comment so that others would not make the same mistake I did.
Sorry.

~~Quickly, merge this before anyone notices!~~
~~This was literally doing the opposite of what it was saying for the hull and shield (only being true when those were higher than 150% combined)~~
~~On top of that, that part was overwritten by the strength in the system; whereas this personality is advertised to activate when either of those conditions is true.~~

~~This meant that frugal escorts did not work properly, and same thing for the Augmented Archon weapons; you could kill the archon all you wanted, so long as you did so in a ship costing less.~~

### CODE REVIEW

**Comment by Hurleveur** [comment]
Quoting from the wiki: `frugal: does not expend ammunition unless it has lost more than 50% of its shields or its government is outnumbered in the current system.`
It doesn't even mention losing hull but that just makes sense I guess

**Comment by warp-core** [comment]
I don't think this is correct.

Let's step through the code:

`AI::AutoFire()`
If the ship has the pacifist personality, or otherwise `CannotAct()`, exit immediately (no autofiring.)
`bool beFrugal` is declared and initialised. If the ship `IsYours()` and your preference is set for escorts to not use ammo, `beFrugal` is true, otherwise (either, the ship isn't youre, or your preference isn't set for escorts to not use ammo) it is false.

Next, we have the contentious code.
`if(person.IsFrugal() || (ship.IsYours() && escortsAreFrugal && escortsUseAmmo))`
If this expression is false, the previously determined value of `beFrugal` is used.
So, if this ship isn't yours, and doesn't have the `frugal` personality, `beFrugal` will be false, and will not be changed.

When it comes to actually firing the weapons, if `beFrugal` is true, and the weapon being considered has `Ammo()` (note, this does not necessarily mean it consumes ammo, just that it has some ammo associated with it, this is how Archons are prevented from using their really powerful weapons all the times, they have ammo specified, but don't actually consume any) it will be skipped. So, when `beFrugal` is false, a ship will always attempt to fire all its weapons, regardless of if they have ammo or not.
This behaviour is correct for a non-player ship without the `frugal` personality.

In the case of a non-player ship with the `frugal` personality, the expression in the if statement will be true, so we proceed into it:
First: `beFrugal = (ship.Hull() + ship.Shields() > 1.5);`
Strictly speaking, this will make `beFrugal` false even if 25% of both shields and hull have been lost, while the comment indicates at least 50% of either. I think that's fine, perhaps the comment should be changed.
If the ships health is above the threshold, `beFrugal` will be true, meaning that when it comes to firing weapons, the ship will not fire ammo weapons. If it is below the threshold, `beFrugal` will be false and the ship will fire ammo weapons.

This change flips the comparison. So, if health is below the threshold, `beFrugal` becomes true, preventing ammo weapon use. This is incorrect.

```
		auto ait = allyStrength.find(ship.GetGovernment());
		auto eit = enemyStrength.find(ship.GetGovernment());
		if(ait != allyStrength.end() && eit != enemyStrength.end() && ait->second < eit->second)
			beFrugal = false;
```
If ally strength is less than enemy strength, make `beFrugal` false (regardless of what it was before, so even if the health was above the threshold, this could cause the ship to use ammo weapons, this is again as the comments describe.)

This PR changes the if statement to: `if(!beFrugal && ait != allyStrength.end() && eit != enemyStrength.end() && ait->second < eit->second)`
So, if this ship and its allies are outgunned, `beFrugal` will only be set to `false` if it is already false. If it is true, (with this PR, meaning if the ship is below the health threshold, which is incorrect, it should be true if this ship is above the health threshold) this will not change that. Essentially, this change means this if statement and the assignment within it do nothing.

**Comment by warp-core** [comment]
The last commit is actually an improvement: it means that the check if this ship and its allies are outgunned iff this ship is already below no longer being frugal. The earlier change is still incorrect, though.

File: source/AI.cpp [hunk 1]
```diff
@@ -2962,13 +2962,16 @@ void AI::AutoFire(const Ship &ship, FireCommand &command, bool secondary) const
 	bool beFrugal = (ship.IsYours() && !escortsUseAmmo);
 	if(person.IsFrugal() || (ship.IsYours() && escortsAreFrugal && escortsUseAmmo))
 	{
-		// Frugal ships only expend ammunition if they have lost 50% of shields
-		// or hull, or if they are outgunned.
+		// Frugal ships do not expend ammunition except if they have lost 50%
+		// of shields and hull combined, or if they are outgunned.
 		beFrugal = (ship.Hull() + ship.Shields() > 1.5);
```
Comment (warp-core) [inline source/AI.cpp:2967]: This isn't 50% of hull and shields combined.
At best, it is 25% of total health, but that assumes hull and shields are equal.
Perhaps `the sum of the fractions of its hull and shield remaining is below 1.5`.
I'm not entirely sure that the comment actually needs changing. It is difficult to come up with an actually better description of what is happening.

Comment (Hurleveur) [reply source/AI.cpp:2967]: Yeah thats why it confused me in the first place...

Comment (Hurleveur) [reply source/AI.cpp:2967]: How about we just say "a significant portion of hull and shields was lost"?

**Comment by Hurleveur** [comment]
Maybe frugal could be improved?
I think it should be using the Health() function instead of Shields() + Hull(), as ships will low shields compared to the hull will just go in rampage instantly (opposite is true, high shields = nothing happens for a while)

File: source/AI.cpp (hunk 1 above)
Comment (warp-core) [reply source/AI.cpp:2967]: Sure

**Comment by Hurleveur** [comment]
I think I finally managed to make the comment clear.
Sorry for the fake alarm, but at least we get a clarification, code improvement, and small logic improvement out of it, I guess.

File: source/AI.cpp [hunk 2]
```diff
@@ -2962,13 +2962,16 @@ void AI::AutoFire(const Ship &ship, FireCommand &command, bool secondary) const
...
-		// Frugal ships only expend ammunition if they have lost 50% of shields
-		// or hull, or if they are outgunned.
-		beFrugal = (ship.Hull() + ship.Shields() > 1.5);
-		auto ait = allyStrength.find(ship.GetGovernment());
-		auto eit = enemyStrength.find(ship.GetGovernment());
-		if(ait != allyStrength.end() && eit != enemyStrength.end() && ait->second < eit->second)
-			beFrugal = false;
+		// The frugal personality is only active when ships have more than 75% of their total health,
+		// and are not outgunned.
+		beFrugal = ship.Health() > .75;
```
Comment (quyykk) [inline source/AI.cpp:2967]:
```suggestion
		beFrugal = (ship.Health() > .75);
```

**Review by quyykk — APPROVED** [review]

**Pull request merged**